### PR TITLE
Updating with logic for expire token

### PR DESCRIPTION
**Pull request by jboucourt**
# Summary
There are issues when vault token expires mainly for jobs running in temporal. We are updating the lib to check when the token is about to expire and proactively reconnect to obtain a new token. New tokens have a ttl of 32 days, our logic is to reconnect 5 days before expire time.

# Testing
**Check one of the following boxes once you've 1) completed testing and added a description of the test below or 2) added a description for why testing isn't needed.**
- [x] This change has been tested and a description of the test(s) performed has been added in accordance with Clear Street's [testing policy](https://www.notion.so/clearstreet/Change-Management-and-Release-Process-v1-1-497e6fecfe1b46d6bc05d645c53fbb3a?pvs=25#a18d9db0239c4313b4c08490021b65a6).
- [ ] Testing has not been performed and a description of why has been added.

**Description of Testing:**

I have tested locally, creating a policy on vault and executing the logic.

How-to Updating gestalt: https://www.notion.so/clearstreet/Updating-Gestalt-7b09f58cdf4c4344a28b5e059d1dc222

### CODE REVIEW

File: gestalt/vault.py [hunk 1]
```diff
@@ -150,6 +152,19 @@ def get(
...
+        # verify if the token still valid, in case not connect()
+        token_details = self.vault_client.auth.token.lookup_self()
+        if token_details is not None and token_details['data'] is not None:
+            expire_time = datetime.fromisoformat(token_details['data']['expire_time'])
+            threshold = timedelta(days=EXPIRATION_THRESHOLD_DAYS)
+            delta_time = expire_time - datetime.now()
+            if delta_time < threshold:
+                self.connect()
+            else:
+                print("Token still valid for: {delta_time} days")
```
Comment (adisunw) [inline gestalt/vault.py:165]: forgot the `f` in the f string

Comment (adisunw) [reply gestalt/vault.py:165]: it also might be worth tucking this logic into a different method altogether.

File: gestalt/vault.py [hunk 2]
```diff
@@ -213,23 +228,20 @@ def _set_secrets_ttl(self, requested_data: Dict[str, Any],
...
+    def worker(self, kube_token: Tuple) -> None:  # type: ignore
         """
         Worker function to renew lease on expiry
         """
         try:
             while self._run_worker:
-                if not token_queue.empty():
-                    token_type, token_id, token_duration = token = token_queue.get(
-                    )
+                if not kube_token:
```
Comment (adisunw) [inline gestalt/vault.py:237]: i think you meant `if kube_token` ?

File: gestalt/vault.py [hunk 3]
```diff
@@ -39,7 +40,7 @@ def __init__(
         self._scheme: str = scheme
         self._run_worker = True
         self.dynamic_token_queue: Queue[Tuple[str, str, str]] = Queue()
-        self.kubes_token_queue: Queue[Tuple[str, str, str]] = Queue()
+        self.kubes_token: Tuple[str, str, str] = ("", "", "")
```
Comment (adisunw) [inline gestalt/vault.py:43]:
```suggestion
        self.kubes_token: Optional[Tuple[str, str, str]] = None
```

**Review by cbrim-12 — APPROVED** [review]

File: gestalt/vault.py [hunk 4]
```diff
@@ -213,23 +219,20 @@ def _set_secrets_ttl(self, requested_data: Dict[str, Any],
...
         """
         Worker function to renew lease on expiry
         """
         try:
             while self._run_worker:
-                if not token_queue.empty():
-                    token_type, token_id, token_duration = token = token_queue.get(
-                    )
+                if kube_token:
+                    token_type, token_id, token_duration = token = kube_token
```
Comment (adisunw) [inline gestalt/vault.py:229]: i don't think you need this `token` variable anymore.

File: gestalt/vault.py [hunk 5]
```diff
@@ -243,3 +246,41 @@ def worker(self, token_queue: Queue) -> None:  # type: ignore
...
+            else:
+                expire_time = expire_time.astimezone(timezone.utc)
+
+            threshold = timedelta(days=EXPIRATION_THRESHOLD_DAYS)
+            current_time = datetime.now(timezone.utc)
+            delta_time = expire_time - current_time
+
+            if delta_time <= threshold:
+                print(f"Re-auth with vault.")
+                self.connect()
```
Comment (adisunw) [inline gestalt/vault.py:282]: this is going to cause additional threads to be created, right? I think we need to stop our existing threads before creating new ones. This sounds like a memory leak.

File: gestalt/vault.py [hunk 6]
```diff
@@ -243,3 +246,41 @@ def worker(self, token_queue: Queue) -> None:  # type: ignore
...
+                return None
+
+            expire_time = token_details['data']['expire_time']
+            # Validate expire_time is present
+            if expire_time is None:
+                print("Cannot parse expire_time, value is None")
+                return None
+
+            # Use isoparse to correctly parse the datetime string
+            expire_time = isoparse(expire_time)
```
Comment (Temurson) [inline gestalt/vault.py:268]: can there be a parsing error here?

File: requirements.test.txt [hunk 7]
```diff
@@ -12,3 +12,4 @@ retry==0.9.2
 types-retry==0.9.9
 jsonpath-ng==1.5.3
 pytest-asyncio==0.19.0
+python-dateutil>=2.8.0
```
Comment (Temurson) [inline requirements.test.txt:15]: do we not need types-python-dateutil?

**Comment by Temurson** [comment]
If you wrote a How-To about testing Gestalt, can you mention a link to it in this PR? Once you're done, would be great to post it in #infra-team too.

File: gestalt/vault.py (hunk 6 above)
Comment (jboucourt) [reply gestalt/vault.py:270]: Yes, the isoparse lib handle that and raise an exception for it.

File: requirements.test.txt (hunk 7 above)
Comment (jboucourt) [reply requirements.test.txt:15]: I'll double check.

File: gestalt/vault.py (hunk 5 above)
Comment (jboucourt) [reply gestalt/vault.py:284]: Being handled.

File: requirements.test.txt (hunk 7 above)
Comment (jboucourt) [reply requirements.test.txt:15]: I checked and python-dateutil requires types-python-dateutil.

**Review by adisunw — APPROVED** [review]

**Review by Temurson — APPROVED** [review]

File: gestalt/vault.py [hunk 8]
```diff
@@ -12,6 +12,9 @@
 from retry.api import retry_call
 
 from gestalt.provider import Provider
+from dateutil.parser import isoparse
+
+EXPIRATION_THRESHOLD_HOURS = 1
```
Comment (adisunw) [inline gestalt/vault.py:17]: no need for hard coding this i think? you get the expiration time from the response.

File: gestalt/vault.py [hunk 9]
```diff
@@ -1,7 +1,7 @@
 import os
-from datetime import datetime, timedelta
+from datetime import datetime, timedelta, timezone
 from queue import Queue
-from threading import Thread
+from threading import Thread, Event
```
Comment (adisunw) [inline gestalt/vault.py:4]: can remove these imports

File: gestalt/vault.py [hunk 10]
```diff
@@ -55,6 +58,7 @@ def __init__(
 
         self.delay = delay
         self.tries = tries
+        # self.kubernetes_ttl_renew: Optional[Thread] = None
```
Comment (adisunw) [inline gestalt/vault.py:61]: can delete this line

File: gestalt/vault.py [hunk 11]
```diff
@@ -213,33 +209,35 @@ def _set_secrets_ttl(self, requested_data: Dict[str, Any],
...
-                "Gestalt Error: Gestalt couldn't connect to Vault")
-        except Exception as err:
-            raise RuntimeError(f"Gestalt Error: {err}")
-
     @property
     def scheme(self) -> str:
         return self._scheme
+
+    def _validate_token_expiration(self) -> None:
+        # token_details = self.vault_client.auth.token.lookup_self()
```
Comment (adisunw) [inline gestalt/vault.py:217]: delete this line

File: setup.py [hunk 12]
```diff
@@ -12,7 +12,7 @@ def readme():
 
 setup(
     name="gestalt-cfg",
-    version="3.4.2",
+    version="3.4.2beta11",
```
Comment (Temurson) [inline setup.py:15]:
```suggestion
    version="3.4.2",
```
should this be changed to just 3.4.2 before merging?

File: gestalt/vault.py [hunk 13]
```diff
@@ -275,14 +228,16 @@ def _validate_token_expiration(self) -> None:
...
 
-            threshold = timedelta(days=EXPIRATION_THRESHOLD_DAYS)
             current_time = datetime.now(timezone.utc)
-            delta_time = expire_time - current_time
+            # in hours
+            delta_time = (expire_time - current_time).total_seconds() / 3600
 
-            if delta_time <= threshold:
-                print(f"Re-auth with vault.")
+            if delta_time < EXPIRATION_THRESHOLD_HOURS:
```
Comment (Temurson) [inline gestalt/vault.py:235]: why did we change expiration threshold to just 1 hour?

File: setup.py (hunk 12 above)
Comment (jboucourt) [reply setup.py:15]: Yup, missed that one.

File: gestalt/vault.py (hunk 13 above)
Comment (jboucourt) [reply gestalt/vault.py:235]: While testing I found that most of the services have a 3h ttl. I want to reduce the amount of calls we make for re-authenticate for token that have less than 1h expiration time.

File: gestalt/vault.py (hunk 8 above)
Comment (jboucourt) [reply gestalt/vault.py:17]: Yes, but we still need to calculate how much time is left. expire_time - current_time > EXPIRATION_THRESHOLD_HOURS.

File: setup.py (hunk 12 above)
Comment (jboucourt) [reply setup.py:15]: Done

**Review by adisunw — APPROVED** [review]

**Pull request merged**